### PR TITLE
New Account Management System

### PR DESCRIPTION
**Pull request by ShahRustam**

### CODE REVIEW

File: platformio/commands/account/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=unused-argument
+
+PIO_ACCOUNT_URL = "http://account.platormio.org/v1"
```
Comment (ivankravets) [inline platformio/commands/account/__init__.py]: Rename to `ACCOUNT_API_URL`

File: platformio/commands/account/command.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            fg="yellow",
+        )
+    resp = requests.post(
+        PIO_ACCOUNT_LOGIN_URL, json={"username": username, "password": password},
+    )
+    try:
+        helpers.process_login_response(resp)
+    except exception.AccountError as e:
+        if "Invalid user credentials" in e.MESSAGE:
+            raise click.ClickException(e.MESSAGE)
```
Comment (ivankravets) [inline platformio/commands/account/command.py]: What will happen with ClickException in this place https://github.com/platformio/platformio-core/blob/develop/platformio/__main__.py#L106?

File: platformio/commands/account/command.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        return click.secho(
+            "You are already logged in with %s account!"
+            % app.get_state_item("account").get("email"),
+            fg="yellow",
+        )
+    resp = requests.post(
+        PIO_ACCOUNT_LOGIN_URL, json={"username": username, "password": password},
+    )
+    try:
+        helpers.process_login_response(resp)
```
Comment (ivankravets) [inline platformio/commands/account/command.py]: Rename to `finalize_login_response`

File: platformio/commands/account/__init__.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+PIO_ACCOUNT_LOGIN_URL = "%s/account/login" % PIO_ACCOUNT_URL
+PIO_ACCOUNT_REFRESH_TOKEN_URL = "%s/token/refresh" % PIO_ACCOUNT_URL
```
Comment (ivankravets) [inline platformio/commands/account/__init__.py]: No need these lines? We can use directly ACCOUNT_API_URL endpoint

File: platformio/commands/account/helpers.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+import jwt
+import requests
+
+from platformio import app, exception
+from platformio.commands.account import (
+    PIO_ACCOUNT_LOGIN_URL,
+    PIO_ACCOUNT_REFRESH_TOKEN_URL,
+)
+
```
Comment (ivankravets) [inline platformio/commands/account/helpers.py]: Implement auto-retry session for account endpoint. See https://www.peterbe.com/plog/best-practice-with-retries-with-requests.

Also, please read official docs.

The case. Something going with a network, we restart account broker, etc. Normally, we should retry within 15sec

File: platformio/commands/account/helpers.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+def get_user_info_from_token(token):
+    try:
+        data = jwt.decode(token, algorithms="RS256", verify=False)
+        return {
+            "username": data.get("preferred_username"),
+            "email": data.get("email"),
+            "email_verified": data.get("email_verified"),
+        }
+    except:  # pylint:disable=bare-except
+        return {}
```
Comment (ivankravets) [inline platformio/commands/account/helpers.py]: Broker should return these data and need to save them to app.state.

File: setup.py [hunk 7]
```diff
@@ -36,6 +36,7 @@
     "tabulate>=0.8.3,<1",
     "pyelftools>=0.25,<1",
     "marshmallow%s" % (">=2,<3" if PY2 else ">=2"),
+    "PyJWT[crypto]==1.7.1",
```
Comment (ivankravets) [inline setup.py]: Remove dependency.

Before commit, please re-run in tox env 
```
make isort format lint
```

File: platformio/commands/account/client.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        resp_json = response.json()
+        if response.status_code != 200:
+            raise exception.AccountError(resp_json.get("message"))
+        app.set_state_item("account", resp_json)
+        return resp_json
+
+    @staticmethod
+    def _get_refresh_token():
+        auth = app.get_state_item("account", {}).get("auth", {})
+        return auth.get("refresh_token")
```
Comment (ivankravets) [inline platformio/commands/account/client.py]: Static methods are normally public, not private

File: platformio/commands/account/client.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+try:
+    from urllib.parse import urljoin
+except ImportError:
+    from urlparse import urljoin
+
+
+class AccountClient(object):
+    def __init__(
+        self,
+        base_url="http://account.platormio.org/v1/",
```
Comment (ivankravets) [inline platformio/commands/account/client.py]: Please rename it to `api_base_url` and default value should be `https://api.account.platformio.org`.

Then, you need to use `self.api_base_url + "v1/login"`

File: platformio/commands/account/command.py [hunk 10]
```diff
@@ -29,14 +29,28 @@ def cli():
...
-@click.option("-u", "--username")
-def account_register(**kwargs):
-    pioplus_call(sys.argv[1:])
+@click.option("-u", "--username", prompt=True)
+@click.option("-e", "--email", prompt=True)
+@click.option(
+    "-p", "--password", prompt=True, hide_input=True, confirmation_prompt=True
+)
+@click.option("-f", "--first-name", prompt=True, default="")
+@click.option("-l", "--last-name", prompt=True, default="")
```
Comment (ivankravets) [inline platformio/commands/account/command.py:38]: 1) Please remove `-f` and `l` short version of options, for the options only full name
2) Why do we need `default=""`? Why could not it be `None`?

File: platformio/commands/account/command.py [hunk 11]
```diff
@@ -29,14 +29,28 @@ def cli():
...
+def account_register(username, email, password, first_name, last_name):
+    client = AccountClient()
+    try:
+        client.registration(username, email, password, first_name, last_name)
+        return click.secho(
+            "User successfully registered! Please, verify your email address.",
```
Comment (ivankravets) [inline platformio/commands/account/command.py:44]: `An account has been successfully created. Please check your mail to activate your account and verify your email address.`

File: platformio/commands/account/command.py (hunk 10 above)
Comment (ShahRustam) [reply platformio/commands/account/command.py:38]: If we send None value in request to broker - we got the error: 
**WARNING:  Invalid HTTP request received.**

Comment (ivankravets) [reply platformio/commands/account/command.py:38]: What is a logical link between CLI and REST API? I do not see any logic. Each interface is responsible for data validation. So, you need to use `first_name or ""` in that place where you do not need None

Comment (ShahRustam) [reply platformio/commands/account/command.py:38]: if we remove default="" - we need remove prompt=True.
I think it is usefully for user.
зображення
(first_name has default="" and we can press enter to use it)

Comment (ivankravets) [reply platformio/commands/account/command.py:38]: Cool

File: platformio/commands/account/command.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            fg="green",
+        )
+    except exception.AccountAlreadyAuthenticated as e:
+        return click.secho(str(e), fg="yellow",)
+
+
+@cli.command("login", short_help="Log in to PIO Account")
+@click.option("-u", "--username", prompt=True)
```
Comment (ivankravets) [inline platformio/commands/account/command.py:52]: Add `info="Username or e-mail"`

File: platformio/commands/account/command.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+@cli.command("token", short_help="Get or regenerate Authentication Token")
+@click.option("-p", "--password")
+@click.option("--regenerate", is_flag=True)
+@click.option("--json-output", is_flag=True)
+def account_token(**kwargs):
+    pioplus_call(sys.argv[1:])
+
+
+@cli.command("forgot", short_help="Forgot password")
+@click.option("-u", "--username")
```
Comment (ivankravets) [inline platformio/commands/account/command.py:94]: info="Username or e-mail"

File: platformio/commands/account/command.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+
+@click.group("account", short_help="Manage PIO Account")
+def cli():
+    pass
+
+
+@cli.command("register", short_help="Create new PIO Account")
+@click.option("-u", "--username", prompt=True)
+@click.option("-e", "--email", prompt=True)
```
Comment (ivankravets) [inline platformio/commands/account/command.py:33]: add above 
```python
def cmd_validate_email(ctx, param, value):  # pylint: disable=unused-argument
    value = str(value).strip()
    if not re.match(r"^[a-z\d_.+-]+@[a-z\d\-]+\.[a-z\d\-.]+$", value, flags=re.I):
        raise click.BadParameter("Invalid E-Mail address")
    return value
```

then use 

```
@click.option("-e", "--email", prompt=True, , callback=cmd_validate_email)
```

File: platformio/commands/account/command.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    client = AccountClient()
+    try:
+        client.change_password(old_password, new_password)
+        return click.secho("Password successfully changed!", fg="green")
+    except exception.AccountNotAuthenticated as e:
+        return click.secho(str(e), fg="yellow",)
+
+
+@cli.command("token", short_help="Get or regenerate Authentication Token")
+@click.option("-p", "--password")
```
Comment (ivankravets) [inline platformio/commands/account/command.py:86]: , prompt=True, hide_input=True

File: platformio/commands/account/client.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+import requests.adapters
+from requests.packages.urllib3.util.retry import Retry  # pylint:disable=import-error
+
+from platformio import app, exception
+
+
+class AccountClient(object):
+    def __init__(
+        self, api_base_url="https://api.account.platormio.org/", retries=3,
```
Comment (ivankravets) [inline platformio/commands/account/client.py:28]: Fix to api_base_url="https://api.accounts.platormio.org"  accountSSS

File: platformio/commands/account/client.py [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    ):
+        if not api_base_url.endswith("/"):
+            api_base_url += "/"
```
Comment (ivankravets) [inline platformio/commands/account/client.py:31]: Remove at the end instead

File: platformio/commands/account/client.py [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        app.set_state_item("account", result)
+        return result
+
+    def logout(self):
+        try:
+            refresh_token = self.get_refresh_token()
+        except:  # pylint:disable=bare-except
+            raise exception.AccountNotAuthenticated()
+        response = requests.post(
+            self.api_base_url + "v1/logout", data={"refresh_token": refresh_token},
```
Comment (ivankravets) [inline platformio/commands/account/client.py:68]: self.api_base_url + "/v1/l

File: platformio/commands/account/command.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+from platformio.commands.account.client import AccountClient
+from platformio.managers.core import pioplus_call
+
+
+@click.group("account", short_help="Manage PIO Account")
+def cli():
+    pass
+
+
+def cmd_validate_username(
```
Comment (ivankravets) [inline platformio/commands/account/command.py:33]: Refactor to `validate_username(value)`, you can keep click.BadParameter

File: platformio/commands/account/command.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+        raise click.BadParameter(
+            "Invalid password format. "
+            "Password must contain at least 8 characters"
+            " including a number and a lowercase letter"
+        )
+    return value
+
+
+@cli.command("register", short_help="Create new PIO Account")
+@click.option("-u", "--username", prompt=True, callback=cmd_validate_username)
```
Comment (ivankravets) [inline platformio/commands/account/command.py:69]:
```
@click.option("-u", "--username", prompt=True, callback=lambda _, __, value: validate_username(value))
```

File: platformio/commands/account/command.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+    client = AccountClient()
+    try:
+        profile = client.get_profile()
+        new_profile = {}
+        for field in profile:
+            new_profile[field] = click.prompt(
+                field.replace("_", " ").capitalize(), default=profile[field]
+            )
+            if field == "email":
+                cmd_validate_email(value=new_profile[field])
```
Comment (ivankravets) [inline platformio/commands/account/command.py:170]: validate_email(new_profile[field])

File: platformio/commands/account/command.py [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+            )
+        if new_profile["username"] != profile["username"]:
+            try:
+                client.logout()
+            except exception.AccountNotAuthenticated:
+                pass
+            return click.secho(
+                "Profile successfully updated! Please re-login.", fg="green"
+            )
+        return click.secho("Profile successfully updated!", fg="green")
```
Comment (ivankravets) [inline platformio/commands/account/command.py:193]: click.secho("Profile successfully updated!", fg="green")
username_changed = new_profile["username"] != profile["username"]
email_changed = new_profile["email"] != profile["email"]
if not username_changed and not email_changed:
     return

            try:
                client.logout()
            except exception.AccountNotAuthenticated:
                pass

if email_changed:
   return click.secho("Please check your mail to verify your new email address and re-login. ", fg="yellow")
click.secho("Please re-login.", fg="yellow")

File: platformio/commands/account/command.py [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+
+@cli.command("logout", short_help="Log out of PIO Account")
+def account_logout():
+    client = AccountClient()
+    try:
+        client.logout()
+        return click.secho("Successfully logged out!", fg="green")
+    except exception.AccountNotAuthenticated as e:
+        return click.secho(str(e), fg="yellow",)
+
```
Comment (ivankravets) [inline platformio/commands/account/command.py:115]: I can't logout when token is broken. If token is broken or is not valid, JUST REMOVE data from state
```
$ pio account logout
Error: Invalid refresh token
```

**Comment by ivankravets** [comment]
Move Account related exceptions to https://github.com/ShahRustam/platformio-core/blob/develop/platformio/commands/account/exception.py
See https://github.com/ShahRustam/platformio-core/blob/develop/platformio/project/exception.py

File: platformio/commands/account/command.py [hunk 24]
```diff
@@ -194,5 +190,85 @@ def account_update(current_password):
...
+def print_profile(profile):
+    click.echo("Profile:")
+    if profile.get("username"):
+        click.echo("- Username: %s" % profile["username"])
+    if profile.get("email"):
+        click.echo("- Email: %s" % profile["email"])
+    if profile.get("first_name"):
+        click.echo("- First name: %s" % profile["first_name"])
+    if profile.get("last_name"):
+        click.echo("- Last name: %s" % profile["last_name"])
```
Comment (ivankravets) [inline platformio/commands/account/command.py:220]: Please keep original CLI style https://github.com/platformio/platformio-core/blob/develop/platformio/commands/device/command.py#L72
Screen Shot 2020-04-17 at 13 18 49

We also use `tabulate` module https://github.com/platformio/platformio-core/blob/develop/platformio/commands/boards.py#L18

File: platformio/commands/account/command.py [hunk 25]
```diff
@@ -194,5 +190,85 @@ def account_update(current_password):
...
+
+
+def print_packages(packages):
+    click.echo()
+    click.echo("Package%s:" % ("s" if len(packages) > 1 else ""))
```
Comment (ivankravets) [inline platformio/commands/account/command.py:225]: Just print `Packages`, no need to check for plural.

File: platformio/commands/account/command.py [hunk 26]
```diff
@@ -194,5 +190,85 @@ def account_update(current_password):
...
+        if "subscription" in package:
+            expire = datetime.datetime.strptime(
+                (
+                    package["subscription"].get("end_at")
+                    or package["subscription"].get("next_bill_at")
+                ),
+                "%Y-%m-%dT%H:%M:%SZ",
+            ).strftime("%Y-%m-%d")
+        click.echo("  Expire: %s" % expire)
+        permissions = []
```
Comment (ivankravets) [inline platformio/commands/account/command.py:238]: Rename in all places `permission` to `service`.

File: platformio/commands/account/command.py [hunk 27]
```diff
@@ -194,5 +190,85 @@ def account_update(current_password):
...
+        for key in package:
+            if not key.startswith("service.") or package[key].get("title"):
```
Comment (ivankravets) [inline platformio/commands/account/command.py:240]: Why? This is a service but there is no title - it's Ok for us! Means, this is a string. Show it!

File: platformio/commands/account/command.py [hunk 28]
```diff
@@ -194,5 +190,85 @@ def account_update(current_password):
...
+                continue
+            permissions.append(package[key].get("title"))
+        if permissions:
+            click.echo("  Permissions: %s" % ", ".join(permissions))
+
+
+def print_subscriptions(subscriptions):
+    click.echo()
+    click.echo("Subscription%s:" % ("s" if len(subscriptions) > 1 else ""))
```
Comment (ivankravets) [inline platformio/commands/account/command.py:249]: Just Subscriptions

**Pull request merged**